elastic/kibana
# [Cases] Refactor connectors to use the form lib

## Summary

This PR converts all connector forms to use the form lib.

Fixes: https://github.com/elastic/kibana/issues/133467, https://github.com/elastic/kibana/issues/143408, https://github.com/elastic/kibana/issues/146394

## Notable changes
- All connector's hooks were converted to use React Query
- All connector's forms were converted to use the Kibana Form Lib
- The form of fields of the connector and the preview of the fields of the connector got separated
- All connector logic in the case view page moved on its own component
- Improve loading visualization on the first render
- Jira: Issue type is not selected automatically
- Jira: Added validation for the issue type field
- ServiceNow: Subcategory is visible

## Testing scenarios

- Create a case with no connectors. Press "Add connector". The Save button should be disabled
- Create a case (or edit an existing one) with a connector. Press the edit pencil. The Save button should be disabled
- Create a case (or edit an existing one) with a connector. Press the edit pencil. Change a field or a connector. The Save button should not be disabled
- Change various connectors and their fields on a case. The fields should be preserved between changes
- Change various connectors and their fields on a case
- Test the forms of all supported connectors in the create case and the view case page
- Push all supported connectors in the create case and the view case page and verify it is working as expected
- Optional fields can be pushed correctly
- Check that the fields are being reset when changing connectors of the same type (Jira for example)
- Test that connectors work in the create case form when attaching outside of Cases (ML for example)
- Jira: Try to submit without an issue type selected. You should see an error
- ServiceNow: Test that the subcategory is being reset when you change the category.

## Flaky tests

The PR fixes the flaky tests in `x-pack/plugins/cases/public/components/create/form_context.test.tsx`. I run the test multiple times on CI to ensure that the test is not flaky anymore: Specifically:
- Test run 50 times in https://github.com/elastic/kibana/pull/157380/commits/f220cc30cf54f9d2dd6942b3ac296260848bf333
- Test run 55 times in https://github.com/elastic/kibana/pull/157380/commits/6221314b2e4504eaf2f822274fc1e018f76cb1b0
- Test run 50 times in https://github.com/elastic/kibana/pull/157380/commits/91551631ffa79142bfe657092438a4360a19feb2

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
